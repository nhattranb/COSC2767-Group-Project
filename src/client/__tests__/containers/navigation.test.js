/*
 *   RMIT University Vietnam
 *   Course: COSC2767 Systems Deployment and Operations
 *   Semester: 2024C
 *   Assessment: Assignment 2
 *   Author: Le Dinh Ngoc Quynh
 *   ID: s3791159
 *   Created  date: 25 Dec 2024
 *   Last modified: 25 Dec 2024
 */

import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Navigation from '../../app/containers/Navigation';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

// Mock components
jest.mock('../../app/components/Common/Button', () => () => 'Button Component');
jest.mock('../../app/components/Store/CartList', () => () => 'CartList Component');
jest.mock('../../app/containers/Cart', () => () => 'Cart Component');
jest.mock('react-autosuggest', () => () => 'Autosuggest Component');

// Mock actions
jest.mock('../../app/actions', () => ({
  ...jest.requireActual('../../app/actions'),
  fetchStoreBrands: () => () => ({ type: 'FETCH_STORE_BRANDS' }),
  fetchStoreCategories: () => () => ({ type: 'FETCH_STORE_CATEGORIES' })
}));

describe('Navigation component', () => {
  let store;

  const initialState = {
    authentication: {
      authenticated: false
    },
    navigation: {
      isMenuOpen: false,
      isCartOpen: false
    },
    cart: {
      cartItems: []
    },
    account: {
      user: null
    },
    brand: {
      storeBrands: []
    },
    category: {
      storeCategories: []
    },
    product: {
      storeProducts: []
    },
    menu: {
      isMenuOpen: false,
      isBrandOpen: false
    },
    search: {
      suggestions: [],
      searchValue: '',
      isLoading: false
    }
  };

  beforeEach(() => {
    store = mockStore(initialState);
    store.dispatch = jest.fn();
  });

  it('renders correctly for unauthenticated user', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Navigation />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly for authenticated user', () => {
    const authenticatedState = {
      ...initialState,
      authentication: {
        authenticated: true
      },
      account: {
        user: {
          role: 'ROLE_MEMBER'
        }
      }
    };
    store = mockStore(authenticatedState);
    store.dispatch = jest.fn();

    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Navigation />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly for admin user', () => {
    const adminState = {
      ...initialState,
      authentication: {
        authenticated: true
      },
      account: {
        user: {
          role: 'ROLE_ADMIN'
        }
      }
    };
    store = mockStore(adminState);
    store.dispatch = jest.fn();

    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Navigation />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with open menu', () => {
    const openMenuState = {
      ...initialState,
      navigation: {
        ...initialState.navigation,
        isMenuOpen: true
      }
    };
    store = mockStore(openMenuState);
    store.dispatch = jest.fn();

    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Navigation />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
}); 
