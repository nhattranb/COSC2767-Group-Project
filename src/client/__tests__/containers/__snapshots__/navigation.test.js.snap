// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component renders correctly for admin user 1`] = `
<DocumentFragment>
  <header
    class="header fixed-mobile-header"
  >
    <div
      class="header-info"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="text-center d-none d-md-block col-md-4"
          >
            <i
              class="fa fa-rocket"
            />
            <span>
              Fast & Free Shipping
            </span>
          </div>
          <div
            class="text-center d-none d-md-block col-md-4"
          >
            <i
              class="fa fa-credit-card"
            />
            <span>
              Flexible Payment Options
            </span>
          </div>
          <div
            class="text-center d-none d-md-block col-md-4"
          >
            <i
              class="fa fa-phone"
            />
            <span>
              24/7 Customer Support
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="align-items-center top-header row"
      >
        <div
          class="pr-0 col-12 order-1 col-sm-12 order-sm-1 col-md-3 order-md-1 col-lg-3 order-lg-1"
        >
          <div
            class="brand"
          >
            <a
              href="/"
            >
              <h1
                class="logo"
              >
                <img
                  alt="favicon"
                  class="rmit-logo"
                  src="/images/rmit-store-logo.png"
                />
                 
              </h1>
            </a>
          </div>
        </div>
        <div
          class="pt-2 pt-lg-0 col-12 order-4 col-sm-12 order-sm-4 col-md-12 order-md-4 col-lg-5 order-lg-2"
        >
          Autosuggest Component
        </div>
        <div
          class="desktop-hidden col-12 order-2 col-sm-12 order-sm-2 col-md-4 order-md-1 col-lg-5 order-lg-3"
        >
          <div
            class="header-links"
          >
            Button ComponentButton Component
          </div>
        </div>
        <div
          class="col-12 order-2 col-sm-12 order-sm-2 col-md-9 order-md-1 col-lg-4 order-lg-3"
        >
          <nav
            class="mt-1 mt-md-0 navbar navbar-expand-md navbar-light bg-light"
          >
            Button Component
            <ul
              class="navbar-nav"
            >
              <li
                class="nav-item"
              >
                <a
                  class="nav-link"
                  href="/shop"
                >
                  Shop
                </a>
              </li>
              <li
                class="dropdown nav-item"
              >
                <a
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="nav-link"
                  href="#"
                >
                  Welcome
                  <span
                    class="fa fa-chevron-down dropdown-caret"
                  />
                </a>
                <div
                  aria-hidden="true"
                  class="dropdown-menu dropdown-menu-right"
                  role="menu"
                  tabindex="-1"
                >
                  <button
                    class="dropdown-item"
                    role="menuitem"
                    tabindex="0"
                    type="button"
                  >
                    Dashboard
                  </button>
                  <button
                    class="dropdown-item"
                    role="menuitem"
                    tabindex="0"
                    type="button"
                  >
                    Sign Out
                  </button>
                </div>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="hidden-mini-cart"
    >
      <div
        class="mini-cart"
      >
        Cart Component
      </div>
      <div
        class="drawer-backdrop"
      />
    </div>
    <div
      aria-hidden="true"
      class="hidden-mini-menu"
    >
      <div
        class="mini-menu"
      >
        <div
          class="navigation-menu"
        >
          <div
            class="menu-header"
          />
          <div
            class="menu-body"
          >
            <div
              class="container"
            >
              <h3
                class="menu-title text-uppercase"
              >
                Shop By Category
              </h3>
              <nav
                role="navigation"
              >
                <ul
                  class="menu-list"
                />
              </nav>
            </div>
          </div>
        </div>
      </div>
      <div
        class="drawer-backdrop"
      />
    </div>
  </header>
</DocumentFragment>
`;

exports[`Navigation component renders correctly for authenticated user 1`] = `
<DocumentFragment>
  <header
    class="header fixed-mobile-header"
  >
    <div
      class="header-info"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="text-center d-none d-md-block col-md-4"
          >
            <i
              class="fa fa-rocket"
            />
            <span>
              Fast & Free Shipping
            </span>
          </div>
          <div
            class="text-center d-none d-md-block col-md-4"
          >
            <i
              class="fa fa-credit-card"
            />
            <span>
              Flexible Payment Options
            </span>
          </div>
          <div
            class="text-center d-none d-md-block col-md-4"
          >
            <i
              class="fa fa-phone"
            />
            <span>
              24/7 Customer Support
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="align-items-center top-header row"
      >
        <div
          class="pr-0 col-12 order-1 col-sm-12 order-sm-1 col-md-3 order-md-1 col-lg-3 order-lg-1"
        >
          <div
            class="brand"
          >
            <a
              href="/"
            >
              <h1
                class="logo"
              >
                <img
                  alt="favicon"
                  class="rmit-logo"
                  src="/images/rmit-store-logo.png"
                />
                 
              </h1>
            </a>
          </div>
        </div>
        <div
          class="pt-2 pt-lg-0 col-12 order-4 col-sm-12 order-sm-4 col-md-12 order-md-4 col-lg-5 order-lg-2"
        >
          Autosuggest Component
        </div>
        <div
          class="desktop-hidden col-12 order-2 col-sm-12 order-sm-2 col-md-4 order-md-1 col-lg-5 order-lg-3"
        >
          <div
            class="header-links"
          >
            Button ComponentButton Component
          </div>
        </div>
        <div
          class="col-12 order-2 col-sm-12 order-sm-2 col-md-9 order-md-1 col-lg-4 order-lg-3"
        >
          <nav
            class="mt-1 mt-md-0 navbar navbar-expand-md navbar-light bg-light"
          >
            Button Component
            <ul
              class="navbar-nav"
            >
              <li
                class="nav-item"
              >
                <a
                  class="nav-link"
                  href="/shop"
                >
                  Shop
                </a>
              </li>
              <li
                class="dropdown nav-item"
              >
                <a
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="nav-link"
                  href="#"
                >
                  Welcome
                  <span
                    class="fa fa-chevron-down dropdown-caret"
                  />
                </a>
                <div
                  aria-hidden="true"
                  class="dropdown-menu dropdown-menu-right"
                  role="menu"
                  tabindex="-1"
                >
                  <button
                    class="dropdown-item"
                    role="menuitem"
                    tabindex="0"
                    type="button"
                  >
                    Dashboard
                  </button>
                  <button
                    class="dropdown-item"
                    role="menuitem"
                    tabindex="0"
                    type="button"
                  >
                    Sign Out
                  </button>
                </div>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="hidden-mini-cart"
    >
      <div
        class="mini-cart"
      >
        Cart Component
      </div>
      <div
        class="drawer-backdrop"
      />
    </div>
    <div
      aria-hidden="true"
      class="hidden-mini-menu"
    >
      <div
        class="mini-menu"
      >
        <div
          class="navigation-menu"
        >
          <div
            class="menu-header"
          />
          <div
            class="menu-body"
          >
            <div
              class="container"
            >
              <h3
                class="menu-title text-uppercase"
              >
                Shop By Category
              </h3>
              <nav
                role="navigation"
              >
                <ul
                  class="menu-list"
                />
              </nav>
            </div>
          </div>
        </div>
      </div>
      <div
        class="drawer-backdrop"
      />
    </div>
  </header>
</DocumentFragment>
`;

exports[`Navigation component renders correctly for unauthenticated user 1`] = `
<DocumentFragment>
  <header
    class="header fixed-mobile-header"
  >
    <div
      class="header-info"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="text-center d-none d-md-block col-md-4"
          >
            <i
              class="fa fa-rocket"
            />
            <span>
              Fast & Free Shipping
            </span>
          </div>
          <div
            class="text-center d-none d-md-block col-md-4"
          >
            <i
              class="fa fa-credit-card"
            />
            <span>
              Flexible Payment Options
            </span>
          </div>
          <div
            class="text-center d-none d-md-block col-md-4"
          >
            <i
              class="fa fa-phone"
            />
            <span>
              24/7 Customer Support
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="align-items-center top-header row"
      >
        <div
          class="pr-0 col-12 order-1 col-sm-12 order-sm-1 col-md-3 order-md-1 col-lg-3 order-lg-1"
        >
          <div
            class="brand"
          >
            <a
              href="/"
            >
              <h1
                class="logo"
              >
                <img
                  alt="favicon"
                  class="rmit-logo"
                  src="/images/rmit-store-logo.png"
                />
                 
              </h1>
            </a>
          </div>
        </div>
        <div
          class="pt-2 pt-lg-0 col-12 order-4 col-sm-12 order-sm-4 col-md-12 order-md-4 col-lg-5 order-lg-2"
        >
          Autosuggest Component
        </div>
        <div
          class="desktop-hidden col-12 order-2 col-sm-12 order-sm-2 col-md-4 order-md-1 col-lg-5 order-lg-3"
        >
          <div
            class="header-links"
          >
            Button ComponentButton Component
          </div>
        </div>
        <div
          class="col-12 order-2 col-sm-12 order-sm-2 col-md-9 order-md-1 col-lg-4 order-lg-3"
        >
          <nav
            class="mt-1 mt-md-0 navbar navbar-expand-md navbar-light bg-light"
          >
            Button Component
            <ul
              class="navbar-nav"
            >
              <li
                class="nav-item"
              >
                <a
                  class="nav-link"
                  href="/shop"
                >
                  Shop
                </a>
              </li>
              <li
                class="dropdown nav-item"
              >
                <a
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="nav-link"
                  href="#"
                >
                  Welcome!
                  <span
                    class="fa fa-chevron-down dropdown-caret"
                  />
                </a>
                <div
                  aria-hidden="true"
                  class="dropdown-menu dropdown-menu-right"
                  role="menu"
                  tabindex="-1"
                >
                  <button
                    class="dropdown-item"
                    role="menuitem"
                    tabindex="0"
                    type="button"
                  >
                    Login
                  </button>
                  <button
                    class="dropdown-item"
                    role="menuitem"
                    tabindex="0"
                    type="button"
                  >
                    Sign Up
                  </button>
                </div>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="hidden-mini-cart"
    >
      <div
        class="mini-cart"
      >
        Cart Component
      </div>
      <div
        class="drawer-backdrop"
      />
    </div>
    <div
      aria-hidden="true"
      class="hidden-mini-menu"
    >
      <div
        class="mini-menu"
      >
        <div
          class="navigation-menu"
        >
          <div
            class="menu-header"
          />
          <div
            class="menu-body"
          >
            <div
              class="container"
            >
              <h3
                class="menu-title text-uppercase"
              >
                Shop By Category
              </h3>
              <nav
                role="navigation"
              >
                <ul
                  class="menu-list"
                />
              </nav>
            </div>
          </div>
        </div>
      </div>
      <div
        class="drawer-backdrop"
      />
    </div>
  </header>
</DocumentFragment>
`;

exports[`Navigation component renders with open menu 1`] = `
<DocumentFragment>
  <header
    class="header fixed-mobile-header"
  >
    <div
      class="header-info"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="text-center d-none d-md-block col-md-4"
          >
            <i
              class="fa fa-rocket"
            />
            <span>
              Fast & Free Shipping
            </span>
          </div>
          <div
            class="text-center d-none d-md-block col-md-4"
          >
            <i
              class="fa fa-credit-card"
            />
            <span>
              Flexible Payment Options
            </span>
          </div>
          <div
            class="text-center d-none d-md-block col-md-4"
          >
            <i
              class="fa fa-phone"
            />
            <span>
              24/7 Customer Support
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="align-items-center top-header row"
      >
        <div
          class="pr-0 col-12 order-1 col-sm-12 order-sm-1 col-md-3 order-md-1 col-lg-3 order-lg-1"
        >
          <div
            class="brand"
          >
            <a
              href="/"
            >
              <h1
                class="logo"
              >
                <img
                  alt="favicon"
                  class="rmit-logo"
                  src="/images/rmit-store-logo.png"
                />
                 
              </h1>
            </a>
          </div>
        </div>
        <div
          class="pt-2 pt-lg-0 col-12 order-4 col-sm-12 order-sm-4 col-md-12 order-md-4 col-lg-5 order-lg-2"
        >
          Autosuggest Component
        </div>
        <div
          class="desktop-hidden col-12 order-2 col-sm-12 order-sm-2 col-md-4 order-md-1 col-lg-5 order-lg-3"
        >
          <div
            class="header-links"
          >
            Button ComponentButton Component
          </div>
        </div>
        <div
          class="col-12 order-2 col-sm-12 order-sm-2 col-md-9 order-md-1 col-lg-4 order-lg-3"
        >
          <nav
            class="mt-1 mt-md-0 navbar navbar-expand-md navbar-light bg-light"
          >
            Button Component
            <ul
              class="navbar-nav"
            >
              <li
                class="nav-item"
              >
                <a
                  class="nav-link"
                  href="/shop"
                >
                  Shop
                </a>
              </li>
              <li
                class="dropdown nav-item"
              >
                <a
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="nav-link"
                  href="#"
                >
                  Welcome!
                  <span
                    class="fa fa-chevron-down dropdown-caret"
                  />
                </a>
                <div
                  aria-hidden="true"
                  class="dropdown-menu dropdown-menu-right"
                  role="menu"
                  tabindex="-1"
                >
                  <button
                    class="dropdown-item"
                    role="menuitem"
                    tabindex="0"
                    type="button"
                  >
                    Login
                  </button>
                  <button
                    class="dropdown-item"
                    role="menuitem"
                    tabindex="0"
                    type="button"
                  >
                    Sign Up
                  </button>
                </div>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="hidden-mini-cart"
    >
      <div
        class="mini-cart"
      >
        Cart Component
      </div>
      <div
        class="drawer-backdrop"
      />
    </div>
    <div
      aria-hidden="false"
      class="mini-menu-open"
    >
      <div
        class="mini-menu"
      >
        <div
          class="navigation-menu"
        >
          <div
            class="menu-header"
          >
            Button Component
          </div>
          <div
            class="menu-body"
          >
            <div
              class="container"
            >
              <h3
                class="menu-title text-uppercase"
              >
                Shop By Category
              </h3>
              <nav
                role="navigation"
              >
                <ul
                  class="menu-list"
                />
              </nav>
            </div>
          </div>
        </div>
      </div>
      <div
        class="drawer-backdrop dark-overflow"
      />
    </div>
  </header>
</DocumentFragment>
`;
